samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerRestoreUtil.java

```diff
@@ -156,53 +158,90 @@ private static CompletableFuture<Map<TaskName, Checkpoint>> restoreAllTaskInstan
           restoreFuture = taskRestoreManager.restore();
         }
 
-        CompletableFuture<Void> taskRestoreFuture = restoreFuture.handle((res, ex) -> {
+        // Order of the following async operations is critical. They are chained as follows:
+        // 1. check if restore succeeded in the first try -> if it failed with DeletedException, flag restore to be retried.
+        // 2. retry restore with getDeleted if previous step completes and returns true.
+        // 3. Close taskRestoreManager ONLY after 1 and 2 completes - to ensure no inflight restore on taskRestoreManager
+        //    fails with closed taskRestoreManager
+
+        CompletableFuture<Boolean> shouldRetryRestoreFuture =
+            // check if the restore should be retried. If restore returned a DeletedException, this future return true
+            shouldRetryFailedRestore(restoreFuture, startTime, samzaContainerMetrics, taskInstanceName);
+
+        CompletableFuture<Checkpoint> taskCheckpointFuture =
```

Comment:
restoreRetryFuture. Return type is already obvious from the CompletableFuture<Checkpoint>